### ISSUE TITLE
[tx] Implement full parameter optimization

### ISSUE THREAD

**Issue by pcmoritz** [NovaSky-AI/SkyRL#518]
The model class is already designed so if adapter_indices is not passed, the base model is used. We should add a parameter to the engine config, and if it is set, it supports full parameter optimization (probably it can only support one client at a time).

I think the main use case for this would be for somebody who has written all their stuff against the tinker API and now wants to do full parameter optimization. They can just stand tx up on their own hardware and do it.

**Comment by Lucas-Fernandes-Martins** [comment]
Hi, I'm currently taking a stab at this but I'm not too happy with the direction of my current solution. 

Basically, I've created a `full_finetuning` argument for the engine, but, in my understanding, adaptations would have to be made in `_create_loss_and_grad_fn` and `process_optim_step` to make sure the non LoRA parameters are updated. For instance, for `_create_loss_and_grad_fn` a possible solution would be adding (to make sure `jax.value_and_grad` receives `argnums=1` instead of 0):

```
 if self.config.full_finetuning:
            # Differentiate with respect to  non_lora_params
            loss_and_grad_fn = jax.value_and_grad(loss_for_lora, argnums=1, has_aux=True)

            if self.config.enforce_eager:
                # Disable JIT compilation for debugging
                self._loss_and_grad_fn = loss_and_grad_fn
            else:
                # Retrieve the sharding of lora and non_lora params and compute the sharding of inputs and outputs
                lora_shardings = jax.tree.map(
                    lambda spec: jax.NamedSharding(self.mesh, spec), nnx.get_partition_spec(self.lora_params)
                )
                non_lora_shardings = jax.tree.map(
                    lambda spec: jax.NamedSharding(self.mesh, spec), nnx.get_partition_spec(self.non_lora_params)
                )
                replicated = jax.NamedSharding(self.mesh, jax.P(None))
                scalar = jax.NamedSharding(self.mesh, jax.P())
                self._loss_and_grad_fn = jax.jit(
                    loss_and_grad_fn,
                    # One input sharding parameter for each argument of loss_for_lora
                    in_shardings=(lora_shardings, non_lora_shardings) + (replicated,) * 8,
                    # One output sharding parameter for each return value of loss_for_lora
                    out_shardings=((scalar, (replicated, replicated)), non_lora_shardings),
                )

        else: 
         
                #current code, using argnums=0
```


However, adding this if statements to check whether we're doing full finetuning or LoRA would bloat the codebase quite a bit.

So I was wondering if you had something else in mind, or some advice on how you believe this should be implemented. 

Thanks in advance :)

**Comment by pcmoritz** [comment]
This is great, thank for working on this! There is a CLI btw (from before the tinker API was implemented), you can take some inspiration from there on how the full param finetuning can be set up: https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl-tx/tx/run/train.py (I was using it to prototype the code)

Concerning your question: There are two different ways how this could be implemented:

1. Make a completely separate loss function for full param, which doesn't split the parameters into lora_params and non_lora_params (more similar to the train.py file above)
2. Try to unify the loss functions, and e.g. set argnums depending on which mode we are in.

The first way might actually make more sense, since in the full param case, we don't even need to consider the lora_params actually -- if you set adapter_indices=None, the whole LoRA part of the model gets completely bypassed. It will probably be cleaner and easier to see that it is in fact implemented correctly, rather than trying to have one loss function that can do both (and might have subtle bugs).

I don't know what exactly the best way is to structure the code, but intuitively, I would probably try to factor out the loss functions into a separate file and include it into engine.py from there. It is probably best to experiment on how to implement it -- first take a pass and just get it working, you can even edit the code in-place and break the LoRA code path. And then you open a PR with that code and we keep iterating on how the best way to structure it is. Or if you have some good ideas on how to get both working at the same time, you can also try that, but focus on getting something working, and then we can clean it up and restructure it together once we see the best way to do it. I'm happy to help with the refactoring etc :)

**Comment by Lucas-Fernandes-Martins** [comment]
Thank you very much for the advice and encouragement! I've spent some time prototyping and I'm finally close to something functional. I'll work asap on some test scripts to make sure it works as expected and I'll raise the PR with what I have.

Once again, thank you for the guidance :)

**Comment by Lucas-Fernandes-Martins** [comment]
Apologies, took a bit longer than I expected, but my first stab is in the following [PR](https://github.com/NovaSky-AI/SkyRL/pull/611). I'm not entirely satisfied with what I came up with, but since it appears to work I've opened the PR to get feedback from folks. Either way, I'll keep iterating to see if I can make it into a cleaner solution.
